internal/batch/batch_future_test.go

```diff
@@ -0,0 +1,297 @@
+package batch
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"math/rand"
+	"sync"
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/assert"
+
+	"go.uber.org/multierr"
+
+	"go.uber.org/cadence/internal"
+	"go.uber.org/cadence/testsuite"
+)
+
+type batchWorkflowInput struct {
+	Concurrency int
+	TotalSize   int
+}
+
+func batchWorkflow(ctx internal.Context, input batchWorkflowInput) ([]int, error) {
+	factories := make([]func(ctx internal.Context) internal.Future, input.TotalSize)
+	for i := 0; i < input.TotalSize; i++ {
+		i := i
+		factories[i] = func(ctx internal.Context) internal.Future {
+			aCtx := internal.WithActivityOptions(ctx, internal.ActivityOptions{
+				ScheduleToStartTimeout: time.Second * 10,
+				StartToCloseTimeout:    time.Second * 10,
+			})
+			return internal.ExecuteActivity(aCtx, batchActivity, i)
+		}
+	}
+
+	batchFuture, err := NewBatchFuture(ctx, input.Concurrency, factories)
+	if err != nil {
+		return nil, err
+	}
+
+	result := make([]int, input.TotalSize)
+	err = batchFuture.Get(ctx, &result)
+	return result, err
+}
+
+func batchWorkflowUsingFutures(ctx internal.Context, input batchWorkflowInput) ([]int, error) {
+	factories := make([]func(ctx internal.Context) internal.Future, input.TotalSize)
+	for i := 0; i < input.TotalSize; i++ {
+		i := i
+		factories[i] = func(ctx internal.Context) internal.Future {
+			aCtx := internal.WithActivityOptions(ctx, internal.ActivityOptions{
+				ScheduleToStartTimeout: time.Second * 10,
+				StartToCloseTimeout:    time.Second * 10,
+			})
+			return internal.ExecuteActivity(aCtx, batchActivity, i)
+		}
+	}
+
+	batchFuture, err := NewBatchFuture(ctx, input.Concurrency, factories)
+	if err != nil {
+		return nil, err
+	}
+	result := make([]int, input.TotalSize)
+
+	for i, f := range batchFuture.GetFutures() {
+		err = f.Get(ctx, &result[i])
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return result, err
+}
+
+func batchActivity(ctx context.Context, taskID int) (int, error) {
+	select {
+	case <-ctx.Done():
+		return taskID, fmt.Errorf("batch activity %d failed: %w", taskID, ctx.Err())
+	case <-time.After(time.Duration(rand.Int63n(100))*time.Millisecond + 900*time.Millisecond):
```

Comment:
probably worth making this a lot faster tbh, currently the tests take ~22s which seems a bit excessive.

the rest of the slowness in here is probably _mostly_ due to trying to measure wall-clock time instead of simulated time, but I think simulated time might be more relevant - there's `env.Now()` for measuring simulated time.  switching to that and e.g. mocking activities with `.After(rand*time.Second)` would make these essentially instant, but would give you some ability to measure "runs concurrently within batch bounds"

Reply:
added a TODO